File: readersWritersGo/writer.go
package main

import (
	"fmt"
	"sync"

	log "github.com/Sirupsen/logrus"
)

// Writer definition
type Writer struct {
	id       int
	resource *Resource
	wg       *sync.WaitGroup
}

// NewWriter creates a Writer
func NewWriter(id int, resource *Resource, wg *sync.WaitGroup) *Writer {
	return &Writer{id: id, resource: resource, wg: wg}
}

func (w *Writer) write() {
	defer w.wg.Done()

	log.Info("Writer starting ", w.id)

	w.resource.write(w.id, fmt.Sprintf("Content #%d", w.id))

	log.Info("Writer done ", w.id)
}

File: readersWritersJava/src/main/java/Writer.java

public class Writer implements Runnable {

    private final int id;
    private final Resource resource;

    public Writer(int id, Resource resource) {
        this.id = id;
        this.resource = resource;
    }

    @Override
    public void run() {
        try {
            System.out.printf("Writer %d starting\n", id);

            resource.write(id, String.format("Content #%d", id));

            System.out.printf("Writer %d is done\n", id);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }
}